Patternlab/dp 26365 new feedback form

clairesunstudio opened the pull request:
Changed:
  - project: Patternlab
    component: MassFeedback, MassFeedbackSuccess
    description: Implement new design and validation logic, and clean up unused components and variants. (#1700)
    issue: DP-26365
    impact: Minor
Added:
  - project: Patternlab
    component: HelpTip
    description: Add HelpTip organism. (#1700)
    issue: DP-26365
    impact: Minor
  - project: Patternlab
    component: IconClose
    description: Add "close" icon. (#1700)
    issue: DP-26365
    impact: Minor
Fixed:
  - project: Patternlab
    component: InputRadio
    description: Fix input radio styles. (#1700)
    issue: DP-26365
    impact: Minor

-----


## Before:
https://mayflower.digital.mass.gov/patternlab/?p=viewall-organisms-feedback
![Screen Shot 2022-11-22 at 11 56 51 AM](https://user-images.githubusercontent.com/5789411/203375260-28352a94-0e29-45fc-a15d-7e2c887e9f69.png)


## After:
https://mayflower.digital.mass.gov/patternlab/b/patternlab/DP-26365-new-feedback-form/?p=viewall-organisms-feedback
![Screen Shot 2022-11-22 at 11 57 57 AM](https://user-images.githubusercontent.com/5789411/203375217-7773a651-2d06-4300-8473-f094245d7548.png)


---

## Review Notes
1. Default form: https://mayflower.digital.mass.gov/patternlab/b/patternlab/DP-26365-new-feedback-form/?p=organisms-mass-feedback
> This screen shows the feedback form by default on every page. 
> Enter and clear inputs to check its validation logic, see [specs in figma] (https://www.figma.com/file/ehFzAlJHDkq2ToJPLsjocn/Feedback-Redesign?node-id=109%3A561&t=JmkmB0scc971W06S-0)
3. Form expanded with the urgent request alert: https://mayflower.digital.mass.gov/patternlab/b/patternlab/DP-26365-new-feedback-form/?p=organisms-mass-feedback-alert.
> This screen shows the form and help tip expanded, with the urgent request alert checked off. 
4. Thank you screen: https://mayflower.digital.mass.gov/patternlab/b/patternlab/DP-26365-new-feedback-form/?p=organisms-mass-feedback-success
> Note that the content in the thank you screen are fully configurable rich text. The contact link condition will be added as data in openmass.

clairesunstudio committed:
add form heading

clairesunstudio committed:
conditional submit button

clairesunstudio committed:
style radio button

clairesunstudio committed:
cleanup fields

clairesunstudio committed:
wip

clairesunstudio committed:
adjust spacing

clairesunstudio committed:
cleanup unused

clairesunstudio committed:
build success page

clairesunstudio committed:
cleanup unused

clairesunstudio committed:
add feedback with alert variant

clairesunstudio committed:
add helptip component

clairesunstudio committed:
add toggle logic

clairesunstudio committed:
use helptip in feedback

clairesunstudio committed:
style and add close event

clairesunstudio committed:
Merge branch 'develop' into patternlab/DP-26365-new-feedback-form

clairesunstudio committed:
fix default state

clairesunstudio committed:
style adjust and wip

clairesunstudio committed:
style wip

clairesunstudio committed:
wip

clairesunstudio committed:
cleanup

clairesunstudio committed:
start adding validation

clairesunstudio committed:
adjust spacing

clairesunstudio committed:
wip

clairesunstudio committed:
add required error message

clairesunstudio committed:
fix has error style and logic

clairesunstudio committed:
fix radio highlight

clairesunstudio committed:
add keyboard accessibility

clairesunstudio committed:
alert and warning messages styles

clairesunstudio committed:
add validation logic

clairesunstudio committed:
improve validation logic

clairesunstudio committed:
update err msg styles

clairesunstudio committed:
fix space shifting issue

clairesunstudio committed:
refactor

clairesunstudio committed:
add changelog

clairesunstudio committed:
test comment out js

clairesunstudio committed:
fix js path

clairesunstudio committed:
fix file name in git

clairesunstudio committed:
fix radio min-width

clairesunstudio committed:
approve react backstop

clairesunstudio committed:
approve patternlab backstop

clairesunstudio committed:
add help-tip readme

clairesunstudio committed:
clean up and document

clairesunstudio commented:
Not sure why keep seeing the those patternlab backstop failures https://app.circleci.com/pipelines/github/massgov/mayflower/4475/workflows/c72cd22a-e81a-4a1d-ab9b-b694d343f8b7/jobs/50907
These seem like false positives, as I'm seeing the image on the page.
https://mayflower.digital.mass.gov/patternlab/b/patternlab/DP-26365-new-feedback-form/?p=pages-contact-row-examples

clairesunstudio committed:
fix mobile tray

clairesunstudio committed:
fix font size and bp

clairesunstudio committed:
fix js logic

clairesunstudio committed:
add animation

clairesunstudio committed:
fix accessiblity

clairesunstudio committed:
fix logic

clairesunstudio committed:
add padding in help text

clairesunstudio committed:
fix a11y labels

clairesunstudio committed:
Merge branch 'develop' into patternlab/DP-26365-new-feedback-form

clairesunstudio committed:
remove unwanted files

clairesunstudio committed:
Merge branch 'patternlab/DP-26365-new-feedback-form' of github.com:massgov/mayflower into patternlab/DP-26365-new-feedback-form

clairesunstudio committed:
fix backstop

clairesunstudio committed:
approve react backstop changes

clairesunstudio commented:
Changes in this PR impact feedback form on mass.gov https://pr1800-5jp8yiaecdsmotpslwrjkdsdqtgznvmt.tugboatqa.com/topics/higher-education 

Please merge this at the same time as the openness changes are ready.

clairesunstudio commented:
@ygannett can you do an accessibility review on the new feedback form in Mayflower?

ygannett commented:
@clairesunstudio Please see my notes on https://massgov.atlassian.net/browse/DP-26365 for the a11y review.

clairesunstudio committed:
combine success screen and form screen templates

clairesunstudio committed:
restructure success data

clairesunstudio committed:
add data

clairesunstudio committed:
fix spacing

clairesunstudio committed:
cleanup separate success template

clairesunstudio committed:
add textarea char count js

clairesunstudio committed:
fix text area styles

clairesunstudio committed:
wip

clairesunstudio committed:
consolidate styles

clairesunstudio committed:
cleanup and format

clairesunstudio committed:
add submitted condition

clairesunstudio committed:
make showWarnMsg prop global

clairesunstudio committed:
update data

clairesunstudio committed:
cleanup

clairesunstudio committed:
Merge branch 'develop' into patternlab/DP-26365-new-feedback-form

clairesunstudio commented:
@arthurbaghdas, per our last discussion:

1. Check with Joe on the org contact info flag logic - checking the box will display the contact message by default on both the form and the success screen. I added a `showWarnMsg` boolean for this purpose.
2. I combined the form and the success templates into one. Added a `submitted` boolean, this will control which screen to show. 
3. Submission logic: I didn't implement this in mayflower, I think this should be in openmass, as the success page should be rendered post-submission not before. Since there will still need to be JS in openmass to handle the form submission, I think that's the natural place to add that logic. 


I'm still working through some accessiblity fixes, but hopefully this unblocked you for implementing the form logic in openmass. Let me know in the morning if you want to discuss more. thanks!

arthurbaghdas committed:
Adding ID for click scroll

arthurbaghdas committed:
Merge branch 'patternlab/DP-26365-new-feedback-form' of github.com:massgov/mayflower into patternlab/DP-26365-new-feedback-form

arthurbaghdas commented:
Hi @clairesunstudio , another thing that I've found is that the negative text area always have a required attribute set, which means whenever I try to submit positive feedback the form is not submitting.
I think we should set the required attribute based on the user selection and this should be done in the scope of Mayflower front-end Implementation

clairesunstudio commented:
Interesting, it was not working like this before. Let me look into this.

clairesunstudio committed:
check for field display before validating

clairesunstudio committed:
check for required fields at submission

clairesunstudio commented:
@arthurbaghdas, what you mentioned above should be fix now. Can you please check again?

clairesunstudio committed:
fix missing id

clairesunstudio committed:
remove aria-required on fieldset

clairesunstudio committed:
change legend to lable and associate to the field

clairesunstudio committed:
add describedBy prop to textarea

clairesunstudio committed:
associate helptext

clairesunstudio committed:
wip

clairesunstudio committed:
change wrapper span to div

clairesunstudio committed:
change fieldset to div for submit button wrapper

clairesunstudio committed:
Combine .ma_help-tip__label-a11y to button label as visually hidden text

clairesunstudio committed:
cleanup

clairesunstudio committed:
add context to heading

clairesunstudio committed:
make error outline thicker

clairesunstudio committed:
add role alert

clairesunstudio committed:
add alert alt text to icon

clairesunstudio commented:
Addressed all accessibility comments

clairesunstudio commented:
Test in openmass PR https://github.com/massgov/openmass/pull/1785

clairesunstudio committed:
add margins to form on mobile

clairesunstudio committed:
add js to hide feedback button

clairesunstudio committed:
style text sizes and line-height

clairesunstudio committed:
implement logic to hide feedback button when feedback form is in view

ygannett commented on packages/patternlab/styleguide/source/_patterns/03-organisms/feedback/mass-feedback.twig:
```diff
@@ -1,9 +1,10 @@
-  <form class="ma__mass-feedback-form__form" method="post" novalidate action="{{ formAction }}" id="{{ formId }}">
-    <fieldset class="ma_feedback-fieldset feedback-load" role="radiogroup"{% if title.required %} aria-required="true"{% endif %}>
+<div data-nosnippet="true" class="ma__mass-feedback-form" id="feedback">
+  <h2 class="ma__mass-feedback-form__heading">{{ heading }}</h2><span class="ma__visually-hidden"> with your feedback</span>
```
Can you put the `span` in the `h2`?  Then, the content of the span is announced as a part of the heading.  Currently, users need to go to the `span` to hear the content.

ygannett commented on packages/patternlab/styleguide/source/_patterns/03-organisms/by-author/help-tip.twig:
```diff
@@ -0,0 +1,47 @@
+      aria-controls="help-tip-content-{{ helpTip.id }}"
+    >
+      <span>{{ helpTip.textTrigger }}</span>
+      <span class="ma_help-tip__label-a11y ma__visually-hidden">Open tooltip to learn more.</span>
```
When the help tip is set up for `aria-describedby`, the button for the tooltip in the container is not interactive with screen readers although the tooltip container content is announced followed by the button label. Screen reader users don't miss the information even the tooltip is collapsed.  But, the visually hidden text becomes not necessary as they don't need to open the tooltip anymore.  Let's remove L.16.  I'm sorry that I asked you to make this change though. 

I believe that keeping the help tip as `aria-describedby` help ensure SR users receive the information with the proper content flow while keyboard and mouse users can click the button to display the tooltip content.

clairesunstudio replied:
Are you suggesting adding `aria-describedby` instead? So it will always announce the helptext using screen readers?

ygannett commented on packages/patternlab/styleguide/source/_patterns/03-organisms/by-author/help-tip.twig:
```diff
@@ -0,0 +1,47 @@
+  >
+    <button
+      class="ma__help-tip__close-mobile"
+      aria-label="close tooltip"
```
Based on above comment, this button doesn't make sense to screen reader users.  Let's remove L.30 aria-label.  It can be commented out for the future when interaction within aria-describedby content becomes available.

ygannett commented on packages/patternlab/styleguide/source/_patterns/03-organisms/by-author/help-tip.twig:
```diff
@@ -0,0 +1,47 @@
+    </button>
+    <button
+      class="ma__help-tip__close-desktop"
+      aria-label="close tooltip"
```
L.36.  Same as above comment.

ygannett reviewed (commented):
Thank you for all the changes.  We have a couple more things to work on. 

When the radio button is clicked, that triggers to display the text box for its associated option.  This is my guess, but it seems the display of the box is considered as a change to trigger the alert and resulting screen readers announce the character count while focus is still on the radio button, which doesn't make sense to SR users.
This is not happening with the current prod. feedback module.  Both markups are set up in the same way.  Not quite sure what's causing this issue.
<img width="679" alt="Array" src="https://user-images.githubusercontent.com/9633303/207374659-d3ffa269-ce57-4cb4-bcfd-35c3c2d08b41.png">

One way to resolve this could be adding `aria-hidden="true"` to `span.remainingChar` when appending it in /assets/js/modules/formInputs.js just to ensure the span is not visible to screen readers, then as the text area gets focus remove the `aria-hidden="true"`.  Any other solutions you think would be more efficient is welcome!

ygannett commented on packages/patternlab/styleguide/source/_patterns/03-organisms/feedback/mass-feedback.twig:
```diff
@@ -13,78 +14,75 @@
 
     <div class="ma__mass-feedback-form__fields">
       <fieldset class="ma_feedback-fieldset">
```
Let's remove the `fieldset` since it's not necessary in this case as we've replaced the `legend` with `label`.

clairesunstudio committed:
add logic to prevent helptext scroll back past the trigger text

clairesunstudio committed:
cleanup console logs

clairesunstudio committed:
cleanup console logs

ygannett commented:
This error message is displayed by hitting the submit button. 
<img width="747" alt="Array" src="https://user-images.githubusercontent.com/9633303/207460064-e9dc86d0-1f1b-46c7-b801-1102a4648d92.png">
I'm afraid that I missed this one at the last review.  Screen reader doesn't recognize it, just because it's displayed above the button after hitting it.  

To make the error message reactive like the other error message, it needs `role="region"`, `aria-live="polite"`, and its ID, also added to the `textarea`'s `aria-controls`.

I think we should think about how to handle this user error:
1. Should we wait to display the message until users hit the button?
2. It's a required field. Is it better to flag it when a user leaves the field empty?(= when the field loses focus after it has focus.)

As the error message is announced, I think it's helpful to set focus on the field, so a user can correct their error on the spot.  To screen reader users, it's not easy to find/locate a field with error to make a correction.

The field is announced as "required", then a user still leaves as blank.  It might be better to remind them it's a required field in the error message.

I'm open for discussion for better solution.  This one is a bit tricky and I see people would think different ways.

clairesunstudio committed:
address comments

clairesunstudio commented:
re: character remaining announcement, I'm getting the opposite experience using VO - the problem you described is in prod mass.gov and not with the new changes. 

re: error message, I think we should keep this experience consistent - it announces when the error shows up, which is when the user hits the submit button. That's usually when required fields are validated, onblur can be too much as they might not be ready to submit yet and it's ok to leave it empty if they don't intend to submit

clairesunstudio committed:
fix and log

clairesunstudio committed:
Merge branch 'develop' into patternlab/DP-26365-new-feedback-form

clairesunstudio committed:
Merge branch 'develop' into patternlab/DP-26365-new-feedback-form

clairesunstudio committed:
add feedback form to bio page

clairesunstudio committed:
remove feedback button from homepage

clairesunstudio committed:
check position

clairesunstudio committed:
fix helptip auto collapse logic

clairesunstudio committed:
fix calculation and cleanup

clairesunstudio committed:
add anchor link

clairesunstudio committed:
fix remaining char announcement on radio button select

clairesunstudio committed:
add isDescription prop to helpTip

clairesunstudio committed:
fix conditional describedby

clairesunstudio committed:
add role alert

clairesunstudio commented:
To handle the error messages appropriately:
>1. To make the error message reactive like the other error message, it needs role="region", aria-live="polite", and its ID, also added to the textarea's aria-controls.

I found that the appropriate role might be "alert", and it's one of the live regions so we don't need to add it explicitly. 

> 2. Add `aria-controls`.
> 3. find the first missed required field and set focus.

clairesunstudio committed:
adjust close button size

clairesunstudio committed:
add aria-controls and ids

clairesunstudio committed:
approve screenshots

clairesunstudio committed:
set focus to the first required field that has an error

clairesunstudio committed:
Merge branch 'develop' into patternlab/DP-26365-new-feedback-form

ygannett commented:
Happened to noticed this duplicate component:
<img width="1448" alt="Array" src="https://user-images.githubusercontent.com/9633303/209235532-1ea386a7-34b3-4574-8686-3cb7dd31a811.png">
It seems that a character count snippet is added whenever the text box gets focus.

clairesunstudio committed:
revert on focus function to fix bug

ygannett commented:
I'm afraid that the contact info. error message's aria-live is cancelled as it positions after the character count message:  https://gaurav5430.medium.com/quick-accessibility-wins-multiple-aria-live-on-single-action-caveat-b79a6f41e7cc
So, to solve this, we could group the char. count message and the error message together in one alert message container.  The container has `aria-live="polite"`, `aria-hidden="true`.  Then, by adding `aria-atomic="true"`, we can ensure all messages are announced.

I tried to avoid this approach by playing around with aria attributes and roles, but nothing works.

Also, we can remove `aria-controls` from `textarea` as the current approach doesn't work.

clairesunstudio committed:
Merge branch 'patternlab/DP-26365-new-feedback-form' of github.com:massgov/mayflower into patternlab/DP-26365-new-feedback-form

clairesunstudio committed:
fix remaining char issue

clairesunstudio committed:
role fixe and default aria-hidden to true

clairesunstudio committed:
Merge branch 'develop' into patternlab/DP-26365-new-feedback-form

arthurbaghdas committed:
Array to string conversion fix

clairesunstudio commented:
@ygannett is going to experiment with the fixes in a follow-up PR.

clairesunstudio committed:
Merge branch 'develop' into patternlab/DP-26365-new-feedback-form

clairesunstudio committed:
remove live char count announcement

clairesunstudio committed:
fix control ids

clairesunstudio committed:
add character count for sr in helptext

clairesunstudio committed:
Merge branch 'patternlab/DP-26365-new-feedback-form' of github.com:massgov/mayflower into patternlab/DP-26365-new-feedback-form

clairesunstudio committed:
Merge branch 'develop' into patternlab/DP-26365-new-feedback-form

clairesunstudio committed:
comment out code that are not used

clairesunstudio committed:
Merge branch 'patternlab/DP-26365-new-feedback-form' of github.com:massgov/mayflower into patternlab/DP-26365-new-feedback-form

clairesunstudio commented:
<img width="1223" alt="Screen Shot 2023-01-10 at 5 35 53 PM" src="https://user-images.githubusercontent.com/5789411/211677023-3becadc7-be9b-4f6f-9349-17e673c67aed.png">

Based on the decision in standup today - I disabled SR announcing the character count live, in favor of the error messages. Making the character count a SR notation as part of the description (help text) in data - this will need to be added in the openmass PR. 

```
    "helpText": "The feedback will only be used for improving the website. If you need assistance, please <a href='mass.gov'>contact the Department of xxxxx</a>. <span class='ma__visually-hidden'>Please limit your input to 500 characters. </span>",
```

tom-fleming committed:
Merge branch 'develop' into patternlab/DP-26365-new-feedback-form

clairesunstudio committed:
Erase text field value when radio button value is changed
